Objects/dictobject.c

```diff
@@ -4791,36 +4864,198 @@ dictiter_iternextitem_lock_held(PyDictObject *d, PyObject *self)
+        // can have either duplicated or skipped attributes, so we strive to do no better
+        // here.
+        int index = get_index_from_order(d, i);
+        PyObject *value = _Py_atomic_load_ptr(&values->values[index]);
+        if (!acquire_key_value(&DK_UNICODE_ENTRIES(k)[index].me_key, value,
+                               &values->values[index], out_key, out_value)) {
+            goto try_locked;
         }
     }
     else {
-        result = PyTuple_New(2);
-        if (result == NULL)
-            return NULL;
-        PyTuple_SET_ITEM(result, 0, Py_NewRef(key));
-        PyTuple_SET_ITEM(result, 1, Py_NewRef(value));
+        Py_ssize_t n = _Py_atomic_load_ssize_relaxed(&k->dk_nentries);
+        if (DK_IS_UNICODE(k)) {
+            PyDictUnicodeEntry *entry_ptr = &DK_UNICODE_ENTRIES(k)[i];
+            PyObject *value;
+            while (i < n &&
+                  (value = _Py_atomic_load_ptr(&entry_ptr->me_value)) == NULL) {
+                entry_ptr++;
+                i++;
+            }
+            if (i >= n)
+                goto fail;
+
+            if (!acquire_key_value(&entry_ptr->me_key, value,
+                                   &entry_ptr->me_value, out_key, out_value)) {
+                goto try_locked;
+            }
+        }
+        else {
+            PyDictKeyEntry *entry_ptr = &DK_ENTRIES(k)[i];
+            PyObject *value;
+            while (i < n &&
+                  (value = _Py_atomic_load_ptr(&entry_ptr->me_value)) == NULL) {
+                entry_ptr++;
+                i++;
+            }
+
+            if (i >= n)
+                goto fail;
+
+            if (!acquire_key_value(&entry_ptr->me_key, value,
+                                   &entry_ptr->me_value, out_key, out_value)) {
+                goto try_locked;
+            }
+        }
     }
-    return result;
+    // We found an element (key), but did not expect it
+    if (_Py_atomic_load_ssize_relaxed(&di->len) == 0) {
+        goto concurrent_modification;
+    }
+    if (!_Py_atomic_compare_exchange_ssize(&di->di_pos, &start_pos, i+1)) {
+        // We lost a race with someone else iterating...
+        if (out_key != NULL) {
+            Py_DECREF(*out_key);
+        }
+        if (out_value != NULL) {
+            Py_DECREF(*out_value);
+        }
+        goto try_locked;
+    }
+    _Py_atomic_add_ssize(&di->len, -1);
+    return 1;
+
+concurrent_modification:
+    PyErr_SetString(PyExc_RuntimeError,
+                    "dictionary keys changed during iteration");
 
 fail:
     di->di_dict = NULL;
     Py_DECREF(d);
-    return NULL;
+    return 0;
+
+    int success;
+try_locked:
+    Py_BEGIN_CRITICAL_SECTION(d);
+    success = dictiter_iternextitem_lock_held(d, self, out_key, out_value);
+    Py_END_CRITICAL_SECTION();
+    return success;
+}
+
+#endif
+
+static bool
+acquire_iter_result(PyObject *result)
+{
+#ifdef Py_GIL_DISABLED
+    if (Py_REFCNT(result) == 1) {
```

Comment:
I'd check:

```
if (_Py_IsOwnedByCurrentThread(result) && Py_REFCNT(result) == 1)
```

And then we don't need the subsequent `Py_REFCNT(result) == 2` check.

`Py_REFCNT()` values can be imprecise if there are concurrent refcounting operations, especially if we are not the owning thread.